File: andre/examples/c/multiply.c
#include <stdio.h>

//unsigned arguments?
int multiply(unsigned int op1, unsigned int op2)
{
    int mask = 0x1;
    int accumulator = 0;

        while(op2) {
            if(mask & op2) {
                accumulator += op1;
            }
            op1 = op1 << 1;
            //if op2 is negative and not defined as unsigned, right shifting shifts 1's into it.
            op2 = op2 >> 1;

            //Try printing the value of op2, to see how this works.
            //printf("%d:\t0x%x\n", op2, op2);
        }

    return accumulator;
}

int main() {

    printf("%d\n", multiply(-1, -15));
    return 0;
}

File: andre/examples/go/multiply.go
package main

import "fmt"

func multiply(a int, b int) int {

    mask := 0x1
    accumulator := 0

    for b != 0 {
        if mask & b == 0x1 {
            accumulator += a
        }

        a = a << 1
        //signed right shifts cause negative numbers to not work
        //Go chose to make the shift behavior depend on whether the shiftee is signed or not
        //In other words, no >>> operator like in Java
        b = int(uint(b) >> 1)

        //uncomment this to learn a bit more about how this works
        //fmt.Printf("%d:\t0x%x\n", b, b);
    }

    return accumulator

}


func main() {
    fmt.Printf("3 * 6 = %d\n", multiply(3, 6))
    fmt.Printf("-3 * 6 = %d\n", multiply(-3, 6))
    fmt.Printf("3 * -6 = %d\n", multiply(3, -6))
    fmt.Printf("1353 * 4799 = %d\n", multiply(1353, 4799))
}